# HG changeset patch
# Parent 5ac5a672256898798f1533c716987a8e46e41be3
Initial commit of right-to-left layout patch

diff -r 5ac5a6722568 lib/codemirror.css
--- a/lib/codemirror.css	Wed May 11 07:17:47 2011 +0200
+++ b/lib/codemirror.css	Sat May 14 01:38:35 2011 +0430
@@ -6,15 +6,15 @@
 }
 
 .CodeMirror-gutter {
-  position: absolute; left: 0; top: 0;
+  position: absolute; right: 0; top: 0;
   background-color: #f7f7f7;
-  border-right: 1px solid #eee;
+  border-left: 1px solid #eee;
   min-width: 2em;
   height: 100%;
 }
 .CodeMirror-gutter-text {
   color: #aaa;
-  text-align: right;
+  text-align: left;
   padding: .4em .2em .4em .4em;
 }
 .CodeMirror-lines {
@@ -38,7 +38,7 @@
   z-index: 10;
   position: absolute;
   visibility: hidden;
-  border-left: 1px solid black !important;
+  border-right: 1px solid black !important;
 }
 .CodeMirror-focused .CodeMirror-cursor {
   visibility: visible;
diff -r 5ac5a6722568 lib/codemirror.js
--- a/lib/codemirror.js	Wed May 11 07:17:47 2011 +0200
+++ b/lib/codemirror.js	Sat May 14 01:38:35 2011 +0430
@@ -150,7 +150,7 @@
       addWidget: function(pos, node, scroll) {
         var pos = localCoords(clipPos(pos), true);
         node.style.top = (showingFrom * lineHeight() + pos.yBot + paddingTop()) + "px";
-        node.style.left = (pos.x + paddingLeft()) + "px";
+        node.style.right = (pos.x + paddingLeft()) + "px";
         code.appendChild(node);
         if (scroll)
           scrollIntoView(pos.x, pos.yBot, pos.x + node.offsetWidth, pos.yBot + node.offsetHeight);
@@ -808,14 +808,14 @@
       while (val.length + pad.length < minwidth) pad += "\u00a0";
       if (pad) firstNode.insertBefore(targetDocument.createTextNode(pad), firstNode.firstChild);
       gutter.style.display = "";
-      lineSpace.style.marginLeft = gutter.offsetWidth + "px";
+      lineSpace.style.marginRight = gutter.offsetWidth + "px";
     }
     function updateCursor() {
       var head = sel.inverted ? sel.from : sel.to;
       var x = charX(head.line, head.ch) + "px", y = (head.line - showingFrom) * lineHeight() + "px";
       inputDiv.style.top = y;
       if (posEq(sel.from, sel.to)) {
-        cursor.style.top = y; cursor.style.left = x;
+        cursor.style.top = y; cursor.style.right = x;
         cursor.style.display = "";
       }
       else cursor.style.display = "none";
@@ -973,7 +973,7 @@
       var visible = options.gutter || options.lineNumbers;
       gutter.style.display = visible ? "" : "none";
       if (visible) updateGutter();
-      else lineDiv.parentNode.style.marginLeft = 0;
+      else lineDiv.parentNode.style.marginRight = 0;
     }
 
     function markText(from, to, className) {
@@ -1122,7 +1122,7 @@
 
       var oldCSS = input.style.cssText;
       input.style.cssText = "position: fixed; width: 30px; height: 30px; top: " + (e.pageY() - 1) +
-        "px; left: " + (e.pageX() - 1) + "px; z-index: 1000; background: white; " +
+        "px; right: " + (e.pageX() - 1) + "px; z-index: 1000; background: white; " +
         "border-width: 0; outline: none; overflow: hidden;";
       var val = input.value = getSelection();
       input.focus();
@@ -1805,9 +1805,9 @@
       else if (this.e.button & 4) return 2;
     },
     pageX: function() {
-      if (this.e.pageX != null) return this.e.pageX;
       var doc = this.target().ownerDocument;
-      return this.e.clientX + doc.body.scrollLeft + doc.documentElement.scrollLeft;
+      if (this.e.pageX != null) return doc.body.clientWidth - this.e.pageX;
+      return doc.body.clientWidth - this.e.clientX - doc.body.scrollLeft - doc.documentElement.scrollLeft;
     },
     pageY: function() {
       if (this.e.pageY != null) return this.e.pageY;
